### PR TITLE
[X86][APX] Avoid generating illegal MI_ND ndd instructions

### PR DESCRIPTION
**Pull request by XinWang10**
The instruction-size limit of 15 bytes still applies to APX instructions. 

Note that it is possible for an EVEX-encoded legacy instruction to reach the 15-byte instruction length limit: 4
bytes of EVEX prefix + 1 byte of opcode + 1 byte of ModRM + 1 byte of SIB + 4 bytes of displacement + 4
bytes of immediate = 15 bytes in total, e.g.

```
addq    $184, -96, %rax   # encoding: [0x62,0xf4,0xfc,0x18,0x81,0x04,0x25,0xa0,0xff,0xff,0xff,0xb8,0x00,0x00,0x00]
```

If we added a segment prefix like fs, the length would be 16.

In such a case, no additional (ASIZE or segment override) prefix can be used. Since this limit is reached only when there is a long immediate, compiler can first load the immediate into a register and then apply the desired prefix(es) to the shorter register-source version of the same instruction class.

### CODE REVIEW

File: llvm/lib/Target/X86/X86InstrInfo.cpp [hunk 1]
```diff
@@ -7251,6 +7251,41 @@ MachineInstr *X86InstrInfo::foldMemoryOperandImpl(
...
+    case X86::XOR64ri32_NF_ND:
+    case X86::OR32ri_ND:
+    case X86::OR64ri32_ND:
+    case X86::OR32ri_NF_ND:
+    case X86::OR64ri32_NF_ND:
+    case X86::ADC32ri_ND:
+    case X86::ADC64ri32_ND:
+    case X86::SBB32ri_ND:
+    case X86::SBB64ri32_ND:
+      return nullptr;
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86InstrInfo.cpp:7286]: Which tests this?

Comment (XinWang10) [reply llvm/lib/Target/X86/X86InstrInfo.cpp:7286]: Not all tested, ADC and SUB tests could test this, they generate MI_ND from mem fold.

File: llvm/test/CodeGen/X86/apx/and.ll [hunk 2]
```diff
@@ -0,0 +1,595 @@
+; NOTE: Assertions have been autogenerated by utils/update_llc_test_checks.py
+; RUN: llc < %s -mtriple=x86_64-unknown -mattr=+ndd -verify-machineinstrs | FileCheck %s
+
+define i8 @and8rr(i8 noundef %a, i8 noundef %b) {
```
Comment (KanRobert) [inline llvm/test/CodeGen/X86/apx/and.ll:4]: Please create a separate PR for this

File: llvm/lib/Target/X86/X86InstrInfo.cpp (hunk 1 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/X86InstrInfo.cpp:7286]: Not got your idea. Do you mean there is no any test for this yet?

File: llvm/test/CodeGen/X86/apx/and.ll (hunk 2 above)
Comment (XinWang10) [reply llvm/test/CodeGen/X86/apx/and.ll:4]: Sure.

File: llvm/lib/Target/X86/X86InstrInfo.cpp (hunk 1 above)
Comment (XinWang10) [reply llvm/lib/Target/X86/X86InstrInfo.cpp:7286]: I mean ADC and SBB test like adc32mi_FS sbb32mi_FS could test this.

File: llvm/lib/Target/X86/X86ISelDAGToDAG.cpp [hunk 3]
```diff
@@ -2923,6 +2926,38 @@ bool X86DAGToDAGISel::selectAddr(SDNode *Parent, SDValue N, SDValue &Base,
...
+                                          SDValue &Index, SDValue &Disp,
+                                          SDValue &Segment) {
+  X86ISelAddressMode AM;
+
+  if (Parent) {
+    unsigned AddrSpace =
+        cast<MemSDNode>(Parent)->getPointerInfo().getAddrSpace();
+    if (AddrSpace == X86AS::GS || AddrSpace == X86AS::FS ||
+        AddrSpace == X86AS::SS)
+      return false;
```
Comment (KanRobert) [inline llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:2943]: Why we not check other segment prefixes here?

File: llvm/lib/Target/X86/X86InstrInfo.cpp (hunk 1 above)
Comment (KanRobert) [reply llvm/lib/Target/X86/X86InstrInfo.cpp:7286]: Need to test all the opcodes here. And add string like "_fold" to function name to indicate that.

Comment (XinWang10) [reply llvm/lib/Target/X86/X86InstrInfo.cpp:7286]: Will do.

File: llvm/lib/Target/X86/X86ISelDAGToDAG.cpp (hunk 3 above)
Comment (XinWang10) [reply llvm/lib/Target/X86/X86ISelDAGToDAG.cpp:2943]: Missing, I didn't find reproducer for others so far, let me try again.

File: llvm/lib/Target/X86/X86InstrArithmetic.td [hunk 4]
```diff
@@ -940,8 +940,8 @@ multiclass ArithBinOp_RFF<bits<8> BaseOpc, bits<8> BaseOpc2, bits<8> BaseOpc4,
     def 32mi8_ND  : BinOpMI8F_RF<mnemonic, Xi32, MemMRM>;
     def 64mi8_ND  : BinOpMI8F_RF<mnemonic, Xi64, MemMRM>;
     def 16mi_ND   : BinOpMIF_RF<0x81, mnemonic, Xi16, opnode, MemMRM>, PD;
-    def 32mi_ND   : BinOpMIF_RF<0x81, mnemonic, Xi32, opnode, MemMRM>;
-    def 64mi32_ND : BinOpMIF_RF<0x81, mnemonic, Xi64, opnode, MemMRM>;
+    def 32mi_ND   : BinOpMIF_RF<0x81, mnemonic, Xi32, opnode, MemMRM, nosegaddr>;
+    def 64mi32_ND : BinOpMIF_RF<0x81, mnemonic, Xi64, opnode, MemMRM, nosegaddr>;
```
[the remark this answers is not in the document]
Comment (XinWang10) [reply llvm/lib/Target/X86/X86InstrArithmetic.td:944]: May add condition for imm size.

**Comment by KanRobert** [comment]
> The instruction-size limit of 15 bytes still applies to APX instructions.
> 
> Note that it is possible for an EVEX-encoded legacy instruction to reach the 15-byte instruction length limit: 4 bytes of EVEX prefix + 1 byte of opcode + 1 byte of ModRM + 1 byte of SIB + 4 bytes of displacement + 4 bytes of immediate = 15 bytes in total, e.g.
> 
> ```
> addq    $184, -96, %rax   # encoding: [0x62,0xf4,0xfc,0x18,0x81,0x04,0x25,0xa0,0xff,0xff,0xff,0xb8,0x00,0x00,0x00]
> ```
> 
> If we added a segment prefix like fs, the length would be 16.
> 
> In such a case, no additional (ASIZE or segment override) prefix can be used. Since this limit is reached only when there is a long immediate, compiler can first load the immediate into a register and then apply the desired prefix(es) to the shorter register-source version of the same instruction class.

@XinWang10 Updated the description for you. In the expand-pseudo solution (discussed offline), you also need to consider the address size prefix.

**Comment by KanRobert** [comment]
Conditional compare/test instructions also have same issues about instruction length limit. I will send a RFC for possible solutions one or two weeks later.

We found this bug when building Node.js with APX features. Such illegal instruction was not emitted when building SPEC17. Maybe we don't need to rush to fix it.

**Comment by KanRobert** [comment]
RFC: https://discourse.llvm.org/t/rfc-support-long-instruction-fixup-for-x86/76539
